Create a lossy index of resource -> ClassPathElement mapping

gsmet opened the pull request:
> [!NOTE]
> This PR is part of a coordinated effort to address the OOM reported in https://github.com/quarkusio/quarkus/issues/42471:
> - https://github.com/quarkusio/quarkus/pull/42492
> - https://github.com/quarkusio/quarkus/pull/42525
> - https://github.com/quarkusio/quarkus/pull/42560

The size of the `ClassLoaderState` of the `QuarkusClassLoader` can be extremely problematic (I have seen it regularly at 10+ MB in the various heap dumps I studied these past months).

I made a first attempt at fixing it a few months ago but wasn't really happy with the approach. I approached the problem differently here.

The idea of this patch is to have a lossy index instead and try to find a good compromise between speed and memory usage.
And that's the only compromise we make as each `ClassPathElement` has a comprehensive index of what it contains so we can narrow down the candidates using the lossy index and then fully check the few ``ClassPathElement``s candidates.

I personally think it makes things a bit more readable too but YMMV.

This is related to issue https://github.com/quarkusio/quarkus/issues/42471 .

In this particular example:
- the old `state` is 13 MB+
- the new index is less than 3 MB

So we end up with 10 less MB per `QuarkusClassLoader` in this case (obviously this will depend on the number of classes in your app).

As for speed, I made a few tests and things are not noticeably slower when starting an application in dev mode from what I could see. I made the implementation of `ClassPathResourceIndex#getResourceKey()` faster in a follow up commit.

One thing that I noticed (but it's preexisting) is that I saw this call appear a few times in the flamegraph:
https://github.com/quarkusio/quarkus/blob/main/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/PathTreeClassPathElement.java#L105-L107

When calling this from the CL we could bypass this check as we know the resource is there. But I'm not sure how to make it evolve from an API perspective. /cc @aloubyansky 

This is better reviewed commit per commit as I added a few more things.

Situation before this patch:

![Screenshot from 2024-08-16 14-17-17](https://github.com/user-attachments/assets/61f5a72b-79d5-414b-a31a-c4e4b63f4b63)

Situation after this patch:

![Screenshot from 2024-08-16 14-15-55](https://github.com/user-attachments/assets/e435d853-475c-4874-92a7-339907cc8c3e)

gsmet committed:
Create a lossy index of resource -> ClassPathElement mapping

The size of ClassLoaderState can be extremely problematic so the idea of
this patch is to have a lossy index instead and try to find a good
compromise between speed and memory usage.

gsmet committed:
Make impl of ClassPathResourceIndex#getResourceKey() faster

Previous implementation:
Warmup Iteration   7: 63202.300 ops/ms
Warmup Iteration   8: 64074.043 ops/ms
Warmup Iteration   9: 63297.222 ops/ms
Warmup Iteration  10: 63905.908 ops/ms

New implementation:
Warmup Iteration   6: 234089.005 ops/ms
Warmup Iteration   7: 234372.742 ops/ms
Warmup Iteration   8: 235722.737 ops/ms
Warmup Iteration   9: 233265.148 ops/ms

Also added a test to make sure we don't break it.

gsmet commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java:
```diff
@@ -183,10 +183,12 @@ private void addCpElement(QuarkusClassLoader.Builder builder, ResolvedDependency
             builder.addParentFirstElement(element);
+            builder.addNormalPriorityElement(element);
         } else if (dep.isFlagSet(DependencyFlags.CLASSLOADER_LESSER_PRIORITY)) {
             builder.addLesserPriorityElement(element);
+        } else {
+            builder.addNormalPriorityElement(element);
         }
-        builder.addElement(element);
```
The semantic have changed a bit here: we used to store all the elements in `elements` and then do some logic to figure out if they were in `lesserPriorityElements` when building the index.
I chose to simplify things and have a clearer semantic.

Thus why the logic has changed here and also why I renamed the method.

gsmet commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java:
```diff
@@ -0,0 +1,241 @@
+ * We try to be clever as to how we build the resource key to reduce the number of misses. It might need further tuning in the
+ * future.
+ */
+public class ClassPathResourceIndex {
```
The removal of `ClassLoaderState` and the addition of this class is the whole purpose of this patch. The rest is really to be able to make this happen.

geoand replied:
It might just be the fact that it's my first day back from PTO, but just by reading the Javadoc, I don't understand how this index is supposed to work. I do understand what problem it tries to solve vs. previous approach, but I don't understand how it does it.

gsmet replied:
It does it by storing a prefix instead of the full class name. A prefix that is deemed to be smart enough to not cause a penalty.
As the `ClassPathElement` contains a list of the resources, we can go through the index to reduce greatly the number of elments to go through and then make sure the resource is there.

gsmet replied:
It's a bit more complex than that as you need to keep full references for some elements (see the sets with local classes, parent first resources...) but that's the general idea of it.

I don't mind adding some javadoc, just tell me if this explanation is good enough for you.

gsmet commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:
```diff
@@ -96,7 +95,11 @@ public static boolean isResourcePresentAtRuntime(String resourcePath) {
     private final String name;
-    private final List<ClassPathElement> elements;
+    // the ClassPathElements to consider are normalPriorityElements + lesserPriorityElements
+    private final List<ClassPathElement> normalPriorityElements;
+    private final List<ClassPathElement> lesserPriorityElements;
```
As explained in https://github.com/quarkusio/quarkus/pull/42525/files#r1719580032, I changed the semantic a bit here.

gsmet committed:
Further simplify QuarkusClassLoader#isApplicationClass()

gsmet committed:
Honor parentFirst in QuarkusClassLoader#getElementsWithResource()

For whatever reason, we weren't honoring parentFirst in this method
which looks like an oversight and un undesirable behavior.

gsmet commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:
```diff
@@ -588,27 +602,30 @@ public List<ClassPathElement> getElementsWithResource(String name) {
 
+        boolean parentFirst = parentFirst(name, getClassPathResourceIndex());
+
         List<ClassPathElement> ret = new ArrayList<>();
-        if (parent instanceof QuarkusClassLoader && !localOnly) {
+
+        if (parentFirst && !localOnly && parent instanceof QuarkusClassLoader) {
             ret.addAll(((QuarkusClassLoader) parent).getElementsWithResource(name));
         }
-        ClassPathElement[] classPathElements = getState().loadableResources.get(name);
-        if (classPathElements == null) {
-            return ret;
+
+        List<ClassPathElement> classPathElements = getClassPathResourceIndex().getClassPathElements(name);
+        ret.addAll(classPathElements);
+
+        if (!parentFirst && !localOnly && parent instanceof QuarkusClassLoader) {
+            ret.addAll(((QuarkusClassLoader) parent).getElementsWithResource(name));
         }
-        ret.addAll(Arrays.asList(classPathElements));
+
         return ret;
```
The behavior of this method was extremely odd as it was always returning the parent resources first. I fixed it to honor parent first in a subsequent commit (hoping it won't break anything - the rest have already been thoroughly tested on CI).

geoand commented:
I'll have a good look at this when I'm back (in 10 days or so)

gsmet commented:
This is ready for review, it passed CI on my fork a few times.

geoand commented on independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:
```diff
@@ -80,6 +80,13 @@ static PathTree ofArchive(Path archive, PathFilter filter) {
         return ArchivePathTree.forPath(archive, filter);
     }
 
+    /**
+     * Whether the PathTree provides local resources.
+     * <p>
+     * For instance a directory or a file provides local resources. A jar does not.
+     */
```
I don't really understand what local resources are

gsmet replied:
Local resources are resources that are local to your deployment. A jar is an external dependency so it's not local.

This is a preexisting notion as the `QuarkusClassLoader` provides the local resources to the tests: see `QuarkusClassLoader#getLocalClassNames()` in the existing code.

geoand replied:
Still sounds confusing to me, but if we are already using that terminology, then 🤷🏽

gsmet replied:
I don't mind changing the name to something that makes more sense to you - I really reused the terminology that was there.
Ideally, let's do that once this patch is in.

aloubyansky replied:
It should be something like `isPathsVisibleOutside()` instead to indicate whether the paths visited while processing this tree are visible outside the context of the `visit` method.

gsmet replied:
Well, that's not the semantic I want here. Because we could make the jar content visible from the outside and that still wouldn't make the jar resources candidate for this. What I want to really know if we consider these classes local to the current deployment.

BTW, I don't mind people bike shedding on the name but please let's do it in a second step. I really want this in.

aloubyansky replied:
> Well, that's not the semantic I want here. Because we could make the jar content visible from the outside and that still wouldn't make the jar resources candidate for this. What I want to really know if we consider these classes local to the current deployment.

From the PathTree perspective that's what this is though. Either the paths are visible or not.

> BTW, I don't mind people bike shedding on the name but please let's do it in a second step. I really want this in.

Sure. That's a public API though, so if it's a follow up it has to go right after this one and if this one gets rolled back, they have to be rolled back together.

gsmet replied:
@aloubyansky I don't think you should place yourself from the PathTree perspective when you define the API: it's really for the consumers that the name is important.

For this very patch, I don't want to know if I'm able to get the content using something else than visit. I want to know if this thing is an archive and thus not something that will be subject to change.
The fact that I have to use `visit` is really an implementation detail (and could change over time) and moreover, that's not what I'm trying to do here: I'm just trying to know if the classes will be subject to change or not.

I'm not saying that my name is ideal but at least from my perspective `isPathsVisibleOutside` has an obscure semantic and I have no idea what I should use it for. And I probably won't have used it if it existed as for me it's not the same semantic.

If you come with a better name, I'll be happy to follow up after this PR is in.

geoand replied:
> thus not something that will be subject to change.

So why not come up with a name that reflects that?

gsmet replied:
I'm all ears but really it can be done in a follow-up as this patch have been around for quite a while now and I want it in.
If you all can come up with a good proposal, I will be happy to do the leg work in another PR.

aloubyansky replied:
> isPathsVisibleOutside has an obscure semantic

Local is a term from a higher (classloading) level. It doesn't apply well to `PathTree` abstraction. `isPathsValidOutside()` is exactly what this is.

aloubyansky replied:
@gsmet is your concern about `OpenContainerPathTree`? Would `isArchiveOrigin()` be better for you?

gsmet replied:
Well, I don't think it is, especially because of the semantic in `MultiRootPathTree` but I pushed the change as I'm a bit tired of arguing.
My personal opinion is that the semantic I chose was good and maybe the name needed some love. Maybe `providesApplicationClasses` to be on par with `QuarkusClassLoader#isApplicationClass()` as it's really the same semantic.
But I don't think the name was that bad once you understand the semantic.

I spent a lot of time on this patch and I would really appreciate if we could take a decision and merge it.

Thanks.

gsmet replied:
Note that if you don't understand the semantic, we might better have a quick call so that I explain it to you.

geoand commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java:
```diff
@@ -0,0 +1,245 @@
+            // for services, we want to reference the full path
+            return resource;
+        }
+        StringBuilder prefix = new StringBuilder();
```
Any chance we can make an okay guess of the length?

gsmet replied:
Not really. And interestingly enough, I made some JMH benchmarks this week-end for SmallRye Config (so totally unrelated) and ended up with the version specifying the length being slightly slower.

Note that I conducted some JMH testing of this method to make sure it's fast (not saying it can't be improved but it's not a naive approach to it - see follow-up commit).

geoand commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:
```diff
@@ -355,46 +333,61 @@ public URL getResource(String unsanitisedName) {
+
+    private static URL getClassPathElementResourceUrl(List<ClassPathElement> classPathElements, String name,
+            boolean endsWithTrailingSlash) {
+        for (ClassPathElement classPathElement : classPathElements) {
```
Can we use a good old `for(i=0...)` here in order to cut down on allocations?

gsmet replied:
Sure we could but I don't think it's worse than it used to be. So ideally, let's do that in a follow-up PR?

geoand replied:
Sure

geoand commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:
```diff
@@ -592,23 +600,20 @@ public List<ClassPathElement> getElementsWithResource(String name, boolean local
         if (parent instanceof QuarkusClassLoader && !localOnly) {
             ret.addAll(((QuarkusClassLoader) parent).getElementsWithResource(name));
         }
-        ClassPathElement[] classPathElements = getState().loadableResources.get(name);
-        if (classPathElements == null) {
+        List<ClassPathElement> classPathElements = getClassPathResourceIndex().getClassPathElements(name);
+        if (classPathElements.isEmpty()) {
             return ret;
         }
-        ret.addAll(Arrays.asList(classPathElements));
+        ret.addAll(classPathElements);
         return ret;
     }
 
-    public List<String> getLocalClassNames() {
+    public Set<String> getLocalClassNames() {
```
Why this change of type?

gsmet replied:
Because that's the semantic we have. A big random bag of unique strings :).

geoand replied:
Yeah, but what do we gain by changing it to a `Set`?

gsmet replied:
The values are coming from Sets now so I decided to be consistent rather than going from a Set to a List, especially since that's the semantic we want.

geoand commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:
```diff
@@ -72,9 +72,16 @@ public static boolean isClassPresentAtRuntime(String className) {
-        return !res.isEmpty();
+        ClassPathResourceIndex classPathResourceIndex = ((QuarkusClassLoader) ccl).getClassPathResourceIndex();
+
+        return classPathResourceIndex.getFirstClassPathElement(resourceName) != null;
```
This can be simplified a tad by using pattern matching with `instanceof`

gsmet replied:
Why not. I don't mind adding some cuteness but I would like to get this patch in and avoid iterating on cosmetic details. I don't think the code is less readable than it used to be.

geoand replied:
Sure, I'm just mentioning it because the whole point of the commit was to simplify the method

geoand reviewed (commented):
This looks like very work, but I do want some clarification on the how the index works before looking even closer :)

geoand commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:
```diff
@@ -239,8 +246,8 @@ public Enumeration<URL> getResources(String unsanitisedName, boolean parentAlrea
         Set<URL> resources = new LinkedHashSet<>();
-        ClassPathElement[] providers = state.loadableResources.get(name);
-        if (providers != null) {
+        List<ClassPathElement> providers = classPathResourceIndex.getClassPathElements(name);
```
This change will also cause an extra allocation the way we iterate over the list

geoand commented:
> I personally think it makes things a bit more readable too but YMMV.

Totally agreed

geoand commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java:
```diff
@@ -0,0 +1,241 @@
+            Map<String, ClassPathElement[]> compactedResourceMapping = new HashMap<>(resourceMapping.size());
+            for (Entry<String, List<ClassPathElement>> resourceMappingEntry : resourceMapping.entrySet()) {
+                compactedResourceMapping.put(resourceMappingEntry.getKey(),
+                        resourceMappingEntry.getValue().toArray(new ClassPathElement[resourceMappingEntry.getValue().size()]));
```
```suggestion
resourceMappingEntry.getValue().toArray(new ClassPathElement[0]));
```

geoand reviewed (approved):
All I really want really at this point is to have the Javadoc of `ClassPathResourceIndex` improved.

Othen than that, all good for me!

gsmet committed:
Remove a hack introduced for Liquibase that is no longer needed.

With the new index, we don't need that anymore. I tested that the test
introduced in this commit is still working fine
(LiquibaseExtensionMigrateAtStartDirectoryChangeLogTest).

Note that I fixed it as I forgot to consider the less priority elements
there and I went for fixing it and then was wondering why this would be
even useful.

I also introduced a shortcut for empty resource.

gsmet commented:
@geoand see the additional commits.

The only change compared to what we discussed is the removal of the `name.isEmpty()` workaround. It's no longer needed and I tested that the Liquibase test you introduced when fixing that is passing properly.
(As mentioned in the commit, I fixed it because I had to fix something there anyway)

Now we just have to pray I didn't break anything with these new changes. I didn't some sanity checks but let's see what CI has to say.

gsmet commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java:
```diff
@@ -117,12 +117,12 @@ public List<ClassPathElement> getClassPathElements(String resource) {
             }
         }
-        return Collections.unmodifiableList(classPathElements);
+        return classPathElements;
```
I think given it's a hot path and not really user-consumable APIs, we could get rid of this safety.

geoand commented:
PR looks good to me.

The additional commits seem like overkill to me, but I don't care enough about that kind of thing to fight it :)

gsmet commented:
@geoand didn’t you ask about these changes this morning? (At least the for loops and the pattern matching)

geoand commented:
Right, what I meant is that having those changes as additional commits seems like overkill to me - my bad for not clarifying

gsmet commented:
Oh I wanted to make it easy for you to review them. I had plans to squash some of the commits (if not too much of a mess). Also I wanted a full CI run green before that to make sure I didn’t break my patch.

geoand commented:
🙏

gsmet commented:
I squashed the commits that were just tiny improvements on top of the rest and kept the ones with a specific semantic that I want to keep around.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathElement.java:
```diff
@@ -74,9 +74,11 @@ default ResolvedDependency getResolvedDependency() {
+     * <p>
+     * This is useful for instance when you want to determine if the resources can be updated in dev mode.
      */
-    boolean providesLocalResources();
```
My comments were about `PathTree` API only. This is a layer on top of it where `PathTree` is an impl detail and so here the methods could reflect the meaning that's relevant in this context.

gsmet committed:
Clarify the API around reloadable classes

Make sure we use the proper vocabulary at each level and make sure the
class loader API is consistently using reloadable instead of local.

Per discussion with Alexey.

gsmet commented:
@aloubyansky I rewrote the last commit as per our discussion. I think we should be all good now but I would appreciate some extra scrutiny. Thanks!
